packages/compass-connections/src/stores/connections-store.ts

```diff
@@ -468,100 +498,72 @@ export function useConnections({
+            connectionOptions: originalConnectionInfo.connectionOptions,
+            defaultAppName: appName,
+            preferences: { forceConnectionOptions, browserCommandForOIDCAuth },
+            notifyDeviceFlow: !isOIDCConnectionAttempt
+              ? undefined
+              : (deviceFlowInformation: {
+                  verificationUrl: string;
+                  userCode: string;
+                }) => {
+                  dispatch({
+                    type: 'oidc-attempt-connect-notify-device-auth',
+                    verificationUrl: deviceFlowInformation.verificationUrl,
+                    userCode: deviceFlowInformation.userCode,
+                  });
+                },
+          }),
+        }
+      );
+
       dispatch({
         type: 'attempt-connect',
+        connectingConnectionId: connectionInfo.id,
         connectingStatusText: `Connecting to ${getConnectionTitle(
           connectionInfo
         )}${
           isOIDCConnectionAttempt
             ? '. Go to the browser to complete authentication.'
             : ''
         }`,
-        connectionAttempt: newConnectionAttempt,
       });
 
       trackConnectionAttemptEvent(connectionInfo);
       debug('connecting with connectionInfo', connectionInfo);
-
-      let notifyDeviceFlow:
-        | ((deviceFlowInformation: {
-            verificationUrl: string;
-            userCode: string;
-          }) => void)
-        | undefined;
-      if (isOIDCConnectionAttempt) {
-        notifyDeviceFlow = (deviceFlowInformation: {
-          verificationUrl: string;
-          userCode: string;
-        }) => {
-          dispatch({
-            type: 'oidc-attempt-connect-notify-device-auth',
-            verificationUrl: deviceFlowInformation.verificationUrl,
-            userCode: deviceFlowInformation.userCode,
-          });
-        };
-      }
-
       log.info(
         mongoLogId(1001000004),
         'Connection UI',
         'Initiating connection attempt'
       );
 
-      const newConnectionDataService = await newConnectionAttempt.connect(
-        adjustConnectionOptionsBeforeConnect({
-          connectionOptions: connectionInfo.connectionOptions,
-          defaultAppName: appName,
-          notifyDeviceFlow,
-          preferences: { forceConnectionOptions, browserCommandForOIDCAuth },
-        })
-      );
-      connectingConnectionAttempt.current = undefined;
-
-      if (!newConnectionDataService || newConnectionAttempt.isClosed()) {
-        // The connection attempt was cancelled.
+      await connectionsManager.connect(adjustedConnectionInfoForConnection);
+      const newConnectionDataService =
+        connectionsManager.getDataServiceForConnection(connectionInfo.id);
+      if (
+        !newConnectionDataService ||
+        connectionsManager.statusOf(connectionInfo.id) !==
+          ConnectionStatus.Connected
+      ) {
+        // The connection attempt was either cancelled or failed
```

Comment:
I'm not sure about communicating failed connection by making `connect` method do nothing when connection fails (due to network error or because the attempt was cancelled) and instead checking for side effects of this methodthat are not really a concern of a connect handling logic. This requires us to spread the state managing logic for this across multiple handles and doesn't give us a single interface that we can rely on for handling connection flow. Is this necessary for some reason?

Reply:
Thanks for raising this and also for additional context around it. I have fixed the guarantees of the connect interface in my recent update.